### ISSUE TITLE
Why not accumulate loss and then take derivative in MAML?

### ISSUE THREAD

**Issue by renesax14** [facebookresearch/higher#60]
Why do you not do this:

```python
def inner_loop2():
    n_inner_iter = 5
    inner_opt = torch.optim.SGD(net.parameters(), lr=1e-1)

    qry_losses = []
    qry_accs = []
    meta_opt.zero_grad()
    meta_loss = 0
    for i in range(task_num):
        with higher.innerloop_ctx(
            net, inner_opt, copy_initial_weights=False
        ) as (fnet, diffopt):
            # Optimize the likelihood of the support set by taking
            # gradient steps w.r.t. the model's parameters.
            # This adapts the model's meta-parameters to the task.
            # higher is able to automatically keep copies of
            # your network's parameters as they are being updated.
            for _ in range(n_inner_iter):
                spt_logits = fnet(x_spt[i])
                spt_loss = F.cross_entropy(spt_logits, y_spt[i])
                diffopt.step(spt_loss)

            # The final set of adapted parameters will induce some
            # final loss and accuracy on the query dataset.
            # These will be used to update the model's meta-parameters.
            qry_logits = fnet(x_qry[i])
            qry_loss = F.cross_entropy(qry_logits, y_qry[i])
            qry_losses.append(qry_loss.detach())
            qry_acc = (qry_logits.argmax(
                dim=1) == y_qry[i]).sum().item() / querysz
            qry_accs.append(qry_acc)

            # Update the model's meta-parameters to optimize the query
            # losses across all of the tasks sampled in this batch.
            # This unrolls through the gradient steps.
            #qry_loss.backward()
            meta_loss += qry_loss

    qry_losses = sum(qry_losses) / task_num
    qry_losses.backward()
    meta_opt.step()
    qry_accs = 100. * sum(qry_accs) / task_num
    i = epoch + float(batch_idx) / n_train_iter
    iter_time = time.time() - start_time
```

instead of what you have:

```python
def inner_loop1():
    n_inner_iter = 5
    inner_opt = torch.optim.SGD(net.parameters(), lr=1e-1)

    qry_losses = []
    qry_accs = []
    meta_opt.zero_grad()
    for i in range(task_num):
        with higher.innerloop_ctx(
            net, inner_opt, copy_initial_weights=False
        ) as (fnet, diffopt):
            # Optimize the likelihood of the support set by taking
            # gradient steps w.r.t. the model's parameters.
            # This adapts the model's meta-parameters to the task.
            # higher is able to automatically keep copies of
            # your network's parameters as they are being updated.
            for _ in range(n_inner_iter):
                spt_logits = fnet(x_spt[i])
                spt_loss = F.cross_entropy(spt_logits, y_spt[i])
                diffopt.step(spt_loss)

            # The final set of adapted parameters will induce some
            # final loss and accuracy on the query dataset.
            # These will be used to update the model's meta-parameters.
            qry_logits = fnet(x_qry[i])
            qry_loss = F.cross_entropy(qry_logits, y_qry[i])
            qry_losses.append(qry_loss.detach())
            qry_acc = (qry_logits.argmax(
                dim=1) == y_qry[i]).sum().item() / querysz
            qry_accs.append(qry_acc)

            # Update the model's meta-parameters to optimize the query
            # losses across all of the tasks sampled in this batch.
            # This unrolls through the gradient steps.
            qry_loss.backward()

    meta_opt.step()
    qry_losses = sum(qry_losses) / task_num
    qry_accs = 100. * sum(qry_accs) / task_num
    i = epoch + float(batch_idx) / n_train_iter
    iter_time = time.time() - start_time
```

https://github.com/facebookresearch/higher/blob/e45c1a059e39a16fa016d37bc15397824c65547c/examples/maml-omniglot.py#L130

-----

https://stackoverflow.com/questions/62394411/why-not-accumulate-query-loss-and-then-take-derivative-in-maml-with-pytorch-and

**Comment by egrefen** [comment]
The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:

```latex
\nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
```

**Comment by renesax14** [comment]
> The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> 
> ```
> \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> ```

I am having this error:

```
Exception has occurred: RuntimeError
Trying to backward through the graph a second time, but the buffers have already been freed. Specify retain_graph=True when calling backward the first time.
```

when tracking higher order grads is True. How did you avoid that?

**Comment by renesax14** [comment]
> The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> 
> ```
> \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> ```

also, why don't you do:

```
meta_loss += qrt_loss.detach()
```

and instead you track them all in a list?

**Comment by renesax14** [comment]
> The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> 
> ```
> \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> ```

Also, why is a task not defined as a single class?

**Comment by egrefen** [comment]
> > The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> > ```
> > \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> > ```
> 
> also, why don't you do:
> 
> ```
> meta_loss += qrt_loss.detach()
> ```
> 
> and instead you track them all in a list?

No particular reason. I guess this allows you to plot task losses separately if you wanted, or measure mean/variance rather than just mean?

**Comment by egrefen** [comment]
> > The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> > ```
> > \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> > ```
> 
> Also, why is a task not defined as a single class?

I'm sorry, I don't understand your question.

**Comment by renesax14** [comment]
> > > The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> > > ```
> > > \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> > > ```
> > 
> > 
> > Also, why is a task not defined as a single class?
> 
> I'm sorry, I don't understand your question.

In meta-learning we sample from a task distribution `p(T)`. I mean why not define `p(T)` as a distribution over the available class labels.

**Comment by renesax14** [comment]
> The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> 
> ```
> \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> ```

I think you still have a mistake because you are not divide by the meta-batch size (even though you are accumulating). This means you'd have to play around the Adam step size more than you need to cuz the quantity of interest is not being normalized correctly.

I suggest to do:

```
qry_loss = F.cross_entropy(qry_logits, y_qry[I]) / task_num
```

instead

**Comment by renesax14** [comment]
> > > The answer to your stack overflow question is the right answer. For any x and theta, `grad(sum(x), theta) = sum(grad(x, theta))`, or in latex:
> > > ```
> > > \nabla_\theta \sum_i f_\theta(x_i) = \sum_i \nabla_\theta f_\theta(x_i)
> > > ```
> > 
> > 
> > also, why don't you do:
> > ```
> > meta_loss += qrt_loss.detach()
> > ```
> > 
> > 
> > and instead you track them all in a list?
> 
> No particular reason. I guess this allows you to plot task losses separately if you wanted, or measure mean/variance rather than just mean?

No, what I mean is why not use `.item()` or `.data`.

Why `.detach()` in particular
